### PR TITLE
Add gui/kitchen-info script - adds more info to the Kitchen screen

### PR DESCRIPTION
**Pull request by CarabusX**

### CODE REVIEW

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:6]:
```suggestion
================
```

Comment (myk002) [inline gui/kitchen-info.lua:6]:
```suggestion
================
```

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:21]: in order for class identity to be preserved across script invocations, classes cannot be declared `local`

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:22]:
```suggestion
kitchen_overlay.ATTRS {
    focus_path='kitchen_overlay',
}
```

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:83]: this can be replaced with a call to `utils.invert()`

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:87]: prefer `#column.label` over `column.label:len()`

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:202]: "Number" is a bit ambiguous. This is the number of items in stock, right? How about calling it "Stock" or "Count"?

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:377]:
```suggestion
    self:sendInputToParent(keys)
```

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:384]: what exactly is this protecting against? does the onStateChange callback fire more than once for the `kitchenpref` screen?

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:391]: instead of checking enabled here, it would be more performance-friendly to remove the onStateChange callback when gui/kitchen-info is disabled.

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:413]:
```suggestion
    print(dfhack.script_help())
```

**Review by myk002 — CHANGES_REQUESTED** [review]

File: gui/kitchen-info.lua
Comment (CarabusX) [reply gui/kitchen-info.lua:21]: Fixed

File: gui/kitchen-info.lua
Comment (CarabusX) [reply gui/kitchen-info.lua:22]: But it appears like this (before change) in many other gui scripts?

File: gui/kitchen-info.lua
Comment (CarabusX) [reply gui/kitchen-info.lua:83]: Done.

File: gui/kitchen-info.lua
Comment (CarabusX) [reply gui/kitchen-info.lua:87]: Fixed here and in 2 other places.

File: gui/kitchen-info.lua
Comment (CarabusX) [reply gui/kitchen-info.lua:202]: But the actual column in DF is labeled "Number".
I will call it "drawNumberColumn".

File: gui/kitchen-info.lua
Comment (CarabusX) [reply gui/kitchen-info.lua]: Will need further change as I still need "parent" for dfhack.screen.isDismissed(parent) check.

File: gui/kitchen-info.lua
Comment (CarabusX) [reply gui/kitchen-info.lua:374]: It protects agains kitchen_overlay reopening when kitchenpref becomes the top screen again after kitchen_overlay is dismissed.
I took the code from "gui/extended-status.lua" script.

I guess it could be replaced with a check of (enabled and not isDismissed(parent))

File: gui/kitchen-info.lua
Comment (CarabusX) [reply gui/kitchen-info.lua]: Done.

File: gui/kitchen-info.lua
Comment (ab9rf) [reply gui/kitchen-info.lua:26]: There's a lot of legacies in our current codebase that are not necessarily worth perpetuating.

Comment (CarabusX) [reply gui/kitchen-info.lua:26]: @myk002 also said: "No functional difference as written, but ATTRS is the official API for setting properties on your class."

Ok, changing it.

File: gui/kitchen-info.lua
Comment (ab9rf) [inline gui/kitchen-info.lua:119]: Is this really the best way to implement this test?

File: gui/kitchen-info.lua
Comment (ab9rf) [inline gui/kitchen-info.lua:149]: What's the significance here of differentiating "flours" from others?

I'm very leery of using substring matches into item names or descriptions; just because it yields the desired results on vanilla items doesn't mean it always will, or that it will do so on modded items as well.

Comment (CarabusX) [reply gui/kitchen-info.lua:149]: The point is user friendlyness.
I am already differenting Dye by the flag.
For the rest when the user sees Flour he will know that by milling Plant Abc he will get Plant Abc Flour, like any other flour in the game.
When he sees Other, he will have to look up what sort of Powder is created and whats its use.

If that doesn't give desired results for some modded Flour, which despite name is not really a Flour, then it is the problem of user-friendlyness of mod that named it.
Yes it is sort of hardcoding vanilla stuff, but most thing on this screen relly on assumptions from vanilla, as many jobs like "Process to bag" or "Mill to paste" are a certain vanilla reaction with certain vanilla tags on items.

File: gui/kitchen-info.lua
Comment (CarabusX) [reply gui/kitchen-info.lua:119]: Again, user experience.
"Leaves" in the name is what the user sees.
LEAVES in growth type, or LEAF material based on LEAF_TEMPLATE material template are all internal tags that can be named whatever and work the same.

(And tbh it is mostly to support vanilla Quarry Bush, which is the only vanilla plant with this reaction. Modded items will just get "Other" so user have to look it up)

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:179]: `dfhack.screen.readTile(x, y)` appears to return nil when ttf mode (F12) is enabled (PRINT_MODE=2D).

Is there a way to handle this?

Comment (CarabusX) [reply gui/kitchen-info.lua:179]: You mean the script crashes now? I admit I didn't test other print modes or ttf.
No idea how much DFHack supports ttf mode, I will have to check.
At least I will try to make the code safer against this.

Comment (CarabusX) [reply gui/kitchen-info.lua]: I added support for detecting and handling TrueType text. It works correctly now.
Also updated changelog, moving it to future release.

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:251]: Would it be possible to detect if `tweak kitchen-prefs-color` is enabled so we could display the corresponding color here?

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:394]: this does not play nice with the search filter. Every time a 'p' appears in the search box, the gui/kitchen-info info is toggled.

I actually think we can get away with not having this option at all. What does toggling add? If a player wants it gone, they can just run `disable gui/kitchen-info`

**Review by myk002 — CHANGES_REQUESTED** [review]
I think this script is a good candidate for "default on". Could you do a PR in the dfhack repo to add it to dfhack.init-example?

**Comment by CarabusX** [comment]
> I think this script is a good candidate for "default on". Could you do a PR in the dfhack repo to add it to dfhack.init-example?

I will happily make a PR to make it enabled by default once it gets merged :)

File: gui/kitchen-info.lua
Comment (myk002) [inline gui/kitchen-info.lua:225]: Are you planning on implementing this? As it stands, this PR looks good except for the P functionality. I'd suggest removing the P toggle for now so we can merge this PR, then we can add it back if and when the search plugin gains the appropriate API.

**Review by myk002 — CHANGES_REQUESTED** [review]

**Review by myk002 — APPROVED** [review]

**Comment by CarabusX** [comment]
@myk002 Thank you for updating it and merging! :)
I got busy with other stuff, and then forgot that it was still open.

**Comment by myk002** [comment]
: ) no worries! It will likely need some updating for v50, if you're interested. I implemented an overlay framework so added information like this won't require its own viewscreen. See: https://docs.dfhack.org/en/latest/docs/dev/overlay-dev-guide.html

**Comment by CarabusX** [comment]
Yeah I would like to have it work in v50.xx as I am playing it at the moment and very much miss the functionality.
But I will probably wait a little until everything settles and stabilizes up a bit after the release.

**Comment by myk002** [comment]
good idea -- I'm still adjusting the DFHack Screen and Gui modules for the new version, and the Lua widget library needs some work as well before it's usable again.

**Pull request merged**
